### ISSUE TITLE
Why do most functions require ArrayBufferView?

### ISSUE THREAD

**Issue by d07RiV** [KhronosGroup/WebGL#2667]
Functions like `bufferData` were extended in WebGL2 to accept offset/length parameters, presumably to avoid having to create temporary views that affect garbage collection, but these functions do not accept `ArrayBuffer` as the data source and instead require a view.

What is the rationale behind this limitation? Is there a technical limitation that makes it more difficult to fetch raw buffer data at a given offset? If I only have an ArrayBuffer to work with, this defeats the purpose of these "optimized" functions as I have to create a view anyway.

For instance, WebAssembly memory only exposes an ArrayBuffer, and since it can dynamically resize (changing the underlying buffer), I would have to do extra comparisons every time to make sure my view is up to date.

**Comment by kenrussell** [comment]
The rationale differs depending on the entry point.

The texture upload functions (for example, texImage2D) require ArrayBufferView because the "GLenum type" argument must match the type of the incoming ArrayBufferView. This restriction was introduced in WebGL 1.0 to reduce the chance of programming errors, and kept during the upgrade to WebGL 2.0. The same constraint applies to readPixels.

When adding the WebGL 2.0 variants to bufferData taking offset and length, since those offsets are in elements and not bytes, it seemed best to only accept ArrayBufferView, since every sub-type defines an element size. There's no technical reason that ArrayBuffer couldn't also be accepted, and the offset treated in bytes.

It would be possible to add overloads to the texImage*, readPixels and other entry points taking ArrayBuffer, and defining slightly relaxed validation rules.

@jdashg any thoughts on adding such overloads to the WebGL 2.0 specification, in support of WebAssembly applications?

**Comment by d07RiV** [comment]
It seems odd to me that the offset-length variants in WebGL2 require a view and an offset expressed in elements instead of bytes, since most use cases I can think of would rathat provide a byte offset in an arbitrary ArrayBuffer - i.e. for both glTF and WebAssembly.

By the way, [this MDN page](https://developer.mozilla.org/en-US/docs/Web/API/WebGL2RenderingContext/getBufferSubData) suggests that ArrayBuffer is accepted in `getBufferSubData`, which is not the case (the provided example doesn't work in Chrome, at least). Did the spec change at some point, or the page simply has a mistake?

**Comment by dmikis** [comment]
I agree with @d07RiV. The [only argument](https://github.com/KhronosGroup/WebGL/pull/1940#issuecomment-238694972) I found for that decision doesn't look convincing to me.

**Comment by RafaelCintron** [comment]
This LGTM.

**Comment by lexaknyazev** [comment]
@kenrussell 
> The texture upload functions (for example, `texImage2D`) require `ArrayBufferView` because the "GLenum type" argument must match the type of the incoming `ArrayBufferView`. ... The same constraint applies to `readPixels`.

> those offsets are in elements and not bytes

Could `readPixels` description in the WebGL 2 spec mention that `dstOffset` is in elements?

(I can see how it's implied but it'd be more convenient to have that explicitly stated.)
